ethers/signer.nim

```diff
@@ -129,18 +95,40 @@ method populateTransaction*(signer: Signer,
   if chainId =? transaction.chainId and chainId != await signer.getChainId():
     raiseSignerError("chain id mismatch")
 
-  var populated = await signer.ensureNonceSequence(transaction)
+  if signer.populateLock.isNil:
+    signer.populateLock = newAsyncLock()
+
+  await signer.populateLock.acquire()
 
-  if populated.sender.isNone:
+  var populated = transaction
+
+  if transaction.sender.isNone:
     populated.sender = some(await signer.getAddress())
-  if populated.chainId.isNone:
+  if transaction.chainId.isNone:
     populated.chainId = some(await signer.getChainId())
-  if populated.gasPrice.isNone and (populated.maxFee.isNone or populated.maxPriorityFee.isNone):
+  if transaction.gasPrice.isNone and (populated.maxFee.isNone or populated.maxPriorityFee.isNone):
     populated.gasPrice = some(await signer.getGasPrice())
-  if populated.nonce.isNone:
-    populated.nonce = some(await signer.getNonce())
-  if populated.gasLimit.isNone:
-          populated.gasLimit = some(await signer.estimateGas(populated))
+
+  if transaction.nonce.isNone and transaction.gasLimit.isNone:
+    # when both nonce and gasLimit are not populated, we must ensure getNonce is
+    # followed by an estimateGas so we can determine if there was an error. If
+    # there is an error, the nonce must be deprecated to prevent nonce gaps and
+    # stuck transactions
+    try:
+      populated.nonce = some(await signer.getNonce())
+      populated.gasLimit = some(await signer.estimateGas(populated))
+    except ProviderError, EstimateGasError:
+      let e = getCurrentException()
+      signer.decreaseNonce()
+      raise e
+    finally:
+      signer.populateLock.release()
```

Comment:
The `try:finally` should be around all the statements following `lock.acquire()`. If one of the `getAddress()` or `getChainId()` or ... fails, then the lock is stuck forever.

Reply:
https://github.com/codex-storage/nim-ethers/pull/58